### DISCUSSION TITLE
Realtime API with Htmx SSE extension

### DISCUSSION THREAD

**Discussion by DnzzL** [pocketbase/pocketbase#4161] [Q&A] [answered]
Hello everyone :wave: 

Has someone managed to use the [Htmx Server-sent-extension ](https://htmx.org/extensions/server-sent-events/) with the Pocketbase realtime API ?
Because on paper, this combo seems smooth.

Note: I'm using Node.

**Comment by ganigeorgiev** [comment, marked as answer]
_Another user recently sent me similar question to the support mail and I'll copy part of my reply here._

If you are establishing the realtime connection to the PocketBase server directly from htmx then I don't think the above will work out of the box because we return JSON reponse and the `sse-swap` as far as I see from the example expect HTML partial (_you'll need also some mechanism to send the PocketBase realtime subscriptions after connecting with the specific client connection id_).

If your realtime flow is `client(eg.browser)<-->Node.js<-->PocketBase` then I guess this will work if you:
- maintain an SSE connection with the SDK between `Node.js<-->PocketBase`
- maintain your own separate SSE connection between the `client(eg.browser)<-->Node.js`
- when your Node server receives a PocketBase event, generate your HTML string and pass it down to the SSE connection between `client<-->Node.js`.

For me personally this type of setup feels a little clunky and without details what you are building I'm not sure I can justify the extra complexity maintaining client state between 2 backends.
Or in other words, if you already have a backend (aka. the Node server) it may not worth your time to integrate PocketBase in your existing stack and you may be better with just Node.

**Reply by DnzzL** [reply]
Ok I see.

Thank you for the answer :+1:

**Reply by RayyanNafees** [reply]
We work with JSON data in HTMX using it's [client-side-templates](https://htmx.org/extensions/client-side-templates/) extension using handlebars (mustache) or nunjucks (jinja2) as template engine

But how do u subscribe to an event of a Pocketbase's Record using htmx?

An example would be highly appreciated

**Reply by ganigeorgiev** [reply]
@RayyanNafees Subscriptions need to be send separately via `POST /api/realtime` (see ["Set subscriptions" api docs](https://pocketbase.io/docs/api-realtime/#set-subscriptions)).

**Reply by RayyanNafees** [reply]
@ganigeorgiev I tried connecting to `POST /api/realtime` and included the `clientId` field in the body (i received from the connection's `PB_CONNECT` event) together with `options` field set to `<collection_name>/<record_id>` but I can't get the record object back
Tried `create`/`update` in action as per the [docs](https://pocketbase.io/docs/api-realtime/#set-subscriptions) but it always returns 204 no content
What's wrong with this request:

```http
POST /api/realtime/ HTTP/1.1

Content-Type: application/json

{
  "clientId": "jsjFfXDZFvvJCInjoe7klGGGXipLw8KeXWX6ojjO",
  "options": "col_name/record_id"
}
```

How can I get the value of the record when it's changed ?

**Reply by ganigeorgiev** [reply]
@RayyanNafees The POST request will just register the subscriptions the specific client is interested in.

Record changes will be received as an event message in the already established long-lived SSE connection (similar to the `PB_CONNECT` event). Note that authorization also happens during the subscription, aka. if you want to authorize as a specific user you'll have to send `Authorization: TOKEN` header with the POST request.

**Reply by ganigeorgiev** [reply]
@RayyanNafees Also I just noticed that you are using `"options"` as parameter but it must be `"subscriptions"`. In any case, if you have further questions please consider opening a new discussion with more details about your use case and ideally some code sample showing how do you interact with the PocketBase APIs.
